### ISSUE TITLE
Integration tests: cucumber-jvm-parallel-plugin not supported

### ISSUE THREAD

**Issue by eharris** [airsonic/airsonic#1240]
The integration tests depend on `cucumber-jvm-parallel-plugin`, but according to the readme at https://github.com/temyers/cucumber-jvm-parallel-plugin that plugin is no longer supported and probably shouldn't be used, and instead upgrading to Cucumber 4.0.0 is recommended.

I wonder if this might be related to whatever has been causing the Travis/CI build unreliability lately.

**Comment by eharris** [comment]
The Cucumber release announcement for 4.0.0 is here: https://cucumber.io/blog/announcing-cucumber-jvm-4-0-0/

**Comment by randomnicode** [comment]
@jvoisin @fxthomas @tesshucom @eharris Here is a PR that gets rid of Cucumber completely and just relies on pure maven failsafe for integration tests: https://github.com/airsonic-advanced/airsonic-advanced/pull/113

My integration test times went from around ~3 mins to ~1 min (factor of 3).

Note that the PR is longer than it needs to be because [airsonic-advanced](https://github.com/airsonic-advanced/airsonic-advanced) does additional testing against external DBs (and thus needs to coordinate and remove some inconsistencies that are in airsonic), but you should be able to just take everything in the `integration-test` folder and make a PR out of it.

**Comment by eharris** [comment]
I would be onboard with doing as @randomnicode suggests and removing cucumber entirely.  But there also seems to be a decent amount of unrelated (or at least not clearly related) stuff that is being touched in the PR he points to.  I vastly prefer well-focused change sets targeted at a single clearly defined objective, as they are much easier for more people to understand and review and verify.

**Comment by randomnicode** [comment]
As mentioned in the comment above:
> you should be able to just take everything in the integration-test folder and make a PR out of it.

That was my suggestion to @fxthomas in another PR also (to just just at the repo folder directly instead of the PR).

The bulk of the touch points in that PR are due to the fact that you don't use external DBs for any testing, which airsonic-advanced does. There are implicit assumptions in the test codebase that don't  work when external dbs are involved and thus require changing. Your integration tests DON'T run against ANY external DB, and thus several underlying issues can be brushed under the rug because they haven't surfaced. airsonic-advanced does and thus it needs to be cognizant of that when setting up its container for the integration tests.

So while you can just say "mvn verify -P integration-tests" now, this invocation will fail miserably when your tests start becoming context dependent, (for instance, the LegacyDatabaseStartupTestCase won't work at all with any external dbs, so you need to exclude it for them). Additionally there things you take for granted when testing with an internal db, the tests often leave the DB in a polluted state, but that's okay because the next test just spins up a new internal db. None of that is possible with an external db. You need to be cognizant of what state your tests are leaving the DB in, because your next test relies on it being "clean". Furthermore you run into edge cases like container defaults not being the same as non-container defaults (for instance the default music folder is different so if you reuse the db between unit and integration tests, you'll have to factor that in).

And so on.

The point is that if you were already running tests against external dbs, you'd necessarily have to make these changes also. But since you don't, then whatever, brush it under the rug, and focus on removing the actual bulk of cucumber.

**Comment by tesshucom** [comment]
I think that the troublesome way of writing tests is one of the reasons why Airsonic's development is slow.
Fine-tuning may be necessary, but I agree with the direction.

**Comment by fxthomas** [comment]
@randomnicode Got a couple of questions if you don't mind (assume ELI5 b/c I don't know a lot about the testing tools in the Java ecosystem):

* If I understand correctly, Surefire and Failsafe both serve as ways to start tests, but Failsafe also adds setup/teardown of external things, right? Any idea why they both exist and why this kind of thing is not handled by the same plugin?
* I want to run only part of the build (here, only the integration tests, skipping the unit tests). I ended up doing a convoluted thing by adding a `surefire.skip` property in the build, but is there a builtin/better way?
* I'm getting issues with `/tmp/music` being created by `root` if it doesn't exist (that's probably the user the Docker daemon is running as), which makes tests run as my local user fail with "permission denied". Creating that folder beforehand works, but did you run into this issue?
* I cannot (yet) get the tests to pass, as they are failing with:
  ```
  [ERROR] Tests run: 3, Failures: 0, Errors: 3, Skipped: 0, Time elapsed: 4.808 s <<< FAILURE! - in org.airsonic.test.StreamIT
  [ERROR] testStreamM4aAsMp3(org.airsonic.test.StreamIT)  Time elapsed: 2.592 s  <<< ERROR!
  java.lang.RuntimeException: no media file id matched
    at org.airsonic.test.StreamIT.lambda$testFileStreaming$2(StreamIT.java:36)
    at org.airsonic.test.StreamIT.testFileStreaming(StreamIT.java:36)
    at org.airsonic.test.StreamIT.testStreamM4aAsMp3(StreamIT.java:19)

  [ERROR] testStreamMp3(org.airsonic.test.StreamIT)  Time elapsed: 1.06 s  <<< ERROR!
  java.lang.RuntimeException: no media file id matched
    at org.airsonic.test.StreamIT.lambda$testFileStreaming$2(StreamIT.java:36)
    at org.airsonic.test.StreamIT.testFileStreaming(StreamIT.java:36)
    at org.airsonic.test.StreamIT.testStreamMp3(StreamIT.java:24)

  [ERROR] testStreamFlacAsMp3(org.airsonic.test.StreamIT)  Time elapsed: 1.082 s  <<< ERROR!
  java.lang.RuntimeException: no media file id matched
    at org.airsonic.test.StreamIT.lambda$testFileStreaming$2(StreamIT.java:36)
    at org.airsonic.test.StreamIT.testFileStreaming(StreamIT.java:36)
    at org.airsonic.test.StreamIT.testStreamFlacAsMp3(StreamIT.java:14)
  ```

  I suspect an assumption that works on your system and/or Travis is not valid on my system, so I'll keep on looking, but do you have any hunch?

**Comment by randomnicode** [comment]
> Got a couple of questions if you don't mind (assume ELI5

Sure. Happy to help. I did not know about the term `ELI5` however. I had to Google that. Now I know.

I'll try and post each answer in a different post so one doesn't get super long

**Comment by randomnicode** [comment]
> If I understand correctly, Surefire and Failsafe both serve as ways to start tests, but Failsafe also adds setup/teardown of external things, right? Any idea why they both exist and why this kind of thing is not handled by the same plugin?

Correct. The plugins serve different purposes and the crucial difference comes from when they're run.

Surefire is for unit tests. It binds to the `test` phase of the lifecycle, which you'll notice is before the `package` phase. That means at that point there isn't even a war/jar or any such packaging. So your unit tests run directly on the compiled classes in their loose individual forms (*.class files usually sitting in `target/generated` or `target/classes` and `target/test-classes`). There isn't a jar/war to test!

Failsafe is for integration tests. It binds to the `integration-test` phase of the lifecycle. This happens after the packaging is done, so now your tests actually run against the full packaged version of your code (that is, airsonic.jar) instead of individual class files. Additionally, you're right that there is a dedicated setup and teardown phase, but the key distinction isn't that they exist, but the idea that they're run _even if the integration-test phase fails_.  The dedicated phases will run regardless and the build only fails afterwards. This isn't true, for instance, say the `test` phase; if your `test` phase fails, maven will err out and not even go to the next (`package`) phase. Finally it is important to note that just running the `integration-test` phase is useless. You need to run `verify` because running the tests by itself doesn't do much. You need to verify whether the tests passed or not.

**Comment by randomnicode** [comment]
> I want to run only part of the build (here, only the integration tests, skipping the unit tests). I ended up doing a convoluted thing by adding a surefire.skip property in the build, but is there a builtin/better way?

I mean you can always specify a given test. So for a unit test, you would just run the test phase but specify which test to run: `mvn -Dtest=org.airsonic.player.TestClass test`. There's more detail on the surefire documentation page: https://maven.apache.org/surefire/maven-surefire-plugin/examples/single-test.html.

Similarly, for an integration test, you would run the verify phase, but specify the integration test to run. However, there is an important caveat. Running the verify phase means the test phase also automatically runs, which means it'll run all your unit tests. So in order to not run those, you can specify a fake non-existent test suite for the unit test. This further adds a slight complication that if it doesn't find a test to run, it'll throw an error, but you can easily tell it to ignore not finding a test for the test phase. So I think the complete invocation is something like `mvn -Dtest=bla -DfailIfNoTests=false -Dit.test=org.airsonic.player.IntegrationTest verify`

**Comment by randomnicode** [comment]
> I'm getting issues with /tmp/music being created by root if it doesn't exist (that's probably the user the Docker daemon is running as), which makes tests run as my local user fail with "permission denied". Creating that folder beforehand works, but did you run into this issue?

I'm not sure how you run it or which command is getting invoked (are you doing it from your IDE?). I do not think this is a Docker daemon thing since the integration tests live outside the docker container (and so does this test directory).

As far as the folder is concerned, you can always override the test directory from the command line. Here's how the GitHub Actions Workflow is set up for the `airsonic-advanced` repo (it does create it first before invoking the maven command) and you can kind of follow it to understand and see where you might be going wrong (workflow file is [here](https://github.com/airsonic-advanced/airsonic-advanced/blob/master/.github/workflows/pr_ci.yml)):
- The workflow runner has the default music directory defined in an env variable: [here](https://github.com/airsonic-advanced/airsonic-advanced/blob/master/.github/workflows/pr_ci.yml#L8)
- This directory is created by the workflow runner: [here](https://github.com/airsonic-advanced/airsonic-advanced/blob/master/.github/workflows/pr_ci.yml#L28)
- It is then passed into Maven for the actual compilation and test invocation as a JVM variable `docker.testing.default-music-folder`: [here](https://github.com/airsonic-advanced/airsonic-advanced/blob/master/.github/workflows/pr_ci.yml#L32) (so you can just specify a different directory in that JVM variable and a different directory will get used)
- This JVM property is actually used in two places by Maven, both in the [integration-test/pom.xml](https://github.com/airsonic-advanced/airsonic-advanced/blob/master/integration-test/pom.xml) 
  - It first binds it to the airsonic container as a volume at /var/music: [here](https://github.com/airsonic-advanced/airsonic-advanced/blob/master/integration-test/pom.xml#L153). remember, that's the default location music location inside airsonic container. So that's good, our default music directory inside the container is pointing to our test music directory outside.
  - The next thing is that we need the integration tests (which run outside the container) to know where the folder is, so they can easily put music in that folder. This is accomplished by adding system variables for the Maven test runners. When generating the environment for its test runners we bind this JVM property to an environment property in the test runner's environment: [here](https://github.com/airsonic-advanced/airsonic-advanced/blob/master/integration-test/pom.xml#L179)
- Finally the tests pick up where the directory is from this environment variable: [here](https://github.com/airsonic-advanced/airsonic-advanced/blob/master/integration-test/src/test/java/org/airsonic/test/Scanner.java#L36)
- The music gets copied to the folder [here](https://github.com/airsonic-advanced/airsonic-advanced/blob/master/integration-test/src/test/java/org/airsonic/test/Scanner.java#L88). This is the only location where the directory gets created if it doesn't exist. Since you're not using a workflow runner, I assume this is where your directory is being created and this is being run as root for you (and it's not in a container). Your base invocation therefore is likely being done as root.

**Comment by randomnicode** [comment]
> I cannot (yet) get the tests to pass, as they are failing with:

Looks like a disconnect with the container.

Your integration tests are putting music in `/tmp/music` in the outside world, but your airsonic-advanced instance that should have:
- /var/music as the default MusicFolder
- have /var/music mapped to /tmp/music as a volume
isn't picking it up.

So check those two assumptions. Possibilities are:
- Your DB is using a different MusicFolder as default (instead of /var/airsonic)
- /var/airsonic isn't mapped to the external volume at /tmp/music

**Comment by randomnicode** [comment]
> These files are relatively huge, and using copyrighted media is probably a bad idea (I assume they are?), I'll see if I can find better samples (e.g. Creative Commons).

These aren't my files. They're actually from airsonic. In the airsonic repo, they're located [here] (https://github.com/airsonic/airsonic/tree/master/integration-test/src/test/resources/blobs/stream) (navigate down the directory tree to see them).

They _are_ very big and I have thought of replacing them, just haven't got around to it yet. It would be great if you found better samples (would need to change a few tests also).

**Comment by randomnicode** [comment]
> Once of the files has a weird encoding, is that normal?

Again, it is from airsonic [here](https://github.com/airsonic/airsonic/tree/master/integration-test/src/test/resources/blobs/stream/dance/input). I think it's by design. I remember seeing a test for weird characters, but don't remember where

**Comment by randomnicode** [comment]
Feel free to ask more questions! Hope this helped!

**Comment by fxthomas** [comment]
That did help a lot. Thanks!

> Surefire is for unit tests. It binds to the `test` phase of the lifecycle, which you'll notice is before the `package` phase. That means at that point there isn't even a war/jar or any such packaging.
> 
> You need to run `verify` because running the tests by itself doesn't do much. You need to verify whether the tests passed or not.

Ah, these two bits helped a lot, that wasn't very clear from a quick read of the official documentation. I think I understand much better now!

> So I think the complete invocation is something like mvn -Dtest=bla -DfailIfNoTests=false -Dit.test=org.airsonic.player.IntegrationTest verify -P integration-test. (EDIT: don't forget the profile that actually adds integration-test module/folder to the build -P integration-test at the end)

Thanks, that sounds a bit hacky but why not if it avoids having to modify the scripts. I'll probably compile a list of useful commands in the "Developer" section of the docs!

> Are you doing it from your IDE?

Nope, straight from the terminal, I usually run `mvn -P integration-test verify` (just like what Travis does).

**Comment by fxthomas** [comment]
More questions:

* Why separate the resource files in different folders (e.g. `integration-test/src/test/resources/blobs/stream/piano/input/piano.mp3`)? Why not just use the Maven "resources" plugin to copy them, or maybe directly bind-mount the container on the source folder?
* Any reason to use `adoptopenjdk` in the Docker container besides the upgrade from JRE8 to JRE11? It's a good thing though, might probably fix some performance issues people are having, so I'll do that separately.

**Comment by fxthomas** [comment]
Okay, got it running with a couple of tweaks, my issue was that I have Docker setup to allow my regular user to run containers, but the Docker daemon was creating non-existing mount points as its own user, i.e. `root`, which prevented the integration tests from writing into them as my regular user.

I'll create PRs when I have that cleaned up!

**Comment by tesshucom** [comment]
> Why separate the resource files in different folders

It's to distinguish lfs managed resources.
Due to the nature of the test, the use of large files is a prerequisite from the start.
For this reason, lfs is used to reduce the burden during regular checkout.
In a normal resource directory, IO slows down.

https://github.com/airsonic/airsonic/blob/master/.gitattributes


Off topic.
lfs history cannot be deleted.
(It's not impossible, but it's not realistic because you need to rehash the entire history.)

There was also a way to manage it on a separate project from the product itself.
One of the irreversible operations.

**Comment by fxthomas** [comment]
Good point re:LFS, I must admit I totally forgot this was enabled here.

**Comment by randomnicode** [comment]
> Any reason to use adoptopenjdk in the Docker container besides the upgrade from JRE8 to JRE11? It's a good thing though, might probably fix some performance issues people are having, so I'll do that separately.

Hmm, the docker image base wasn't changed in that PR. Are you talking in general?
In general, I think it's more practical and feasible to be on the current LTS release, which happens to be 11. If you notice, all the external db tests run on 11 because it the current LTS release. That was a conscious decision. `airsonic-advanced` would move the compilation base to 11 too, but it's a breaking change and thus is waiting for the next major change (11.0 stable will likely be on Java 11) No particular reason otherwise. JDK8 is very old at this point in time.

> Why separate the resource files in different folders (e.g. integration-test/src/test/resources/blobs/stream/piano/input/piano.mp3)? Why not just use the Maven "resources" plugin to copy them, or maybe directly bind-mount the container on the source folder?

I think a large part of that is just borrowing from the historical folder layout in `airsonic`, which as @tesshucom points is possibly a remnant of lfs. However, a key additional point here is future test independence. Mounting directly in the folder is possible, but I wanted the test fixtures independent. Currently, there aren't very many tests, but some day there might be. If some test were to delete a file in its own mounted directory, it's not a big deal, the following test just restores the file from the resources folder (or creates a new folder altogether and copies over the files from resources). This kind of independence is not possible if the resources folder is itself mounted, because deleting a file there means it's lost to every other test in the future (there is no "ground truth").

> I'll create PRs when I have that cleaned up!

Awesome! Perfect! Happy coding!

**Comment by tesshucom** [comment]
> a key additional point here is future test independence.

It is important.
The integration test is often separated from the body to ensure portability.
In many cases, it requires a completely different technology than the main project.

The situation is more severe when using lfs.
Depending on the handling, the free plan will eventually exceed lfs and may not be reproducible.

> because you need to rehash the entire history

The integration test history isn't that important, so if it's separate from the body, recreating the repository is the easiest way to clean up lfs.
(It is absolutely impossible to do with Airsonic today)

Ensuring portability is useful for engineers who want to perform large-scale verification such as randomnicode.

Clone seems to remind some users of the dispersion of engineers, but I don't think so.
For large projects, being able to easily create a sandbox for complex validation can be an advantage.
